File: ProtoSAS/src/main/java/dd/sas/pipeline/calculation/processor/associativeprocessors/AssociativeProcessor.java
package dd.sas.pipeline.calculation.processor.associativeprocessors;

import dd.sas.pipeline.worldmodel.Relation;
import dd.sas.pipeline.worldmodel.WorldObject;

import java.util.function.Supplier;

/**
 * Created by Sergey on 18.09.2016.
 */
public abstract class AssociativeProcessor<D extends WorldObject, R extends WorldObject> {

    public <D1 extends D, R1 extends R, S extends Relation<D1, R1>> S generator(
            Supplier<S> supplier, D1 d, R1 r) {
        S see = supplier.get();
        see.setDomain(d);
        see.setRange(r);
        return see;
    }

}

File: ProtoSASOnScala/src/main/scala/dd/sas/pipeline/calculation/processors/associativeprocessors/AssociativeProcessor.scala
package dd.sas.pipeline.calculation.processors.associativeprocessors

import dd.sas.pipeline.calculation.processors.Processor
import dd.sas.pipeline.worldmodel.{Nihility, Relation, WorldObject}

/**
  * Created by Sergey on 02.10.2016.
  */
abstract class AssociativeProcessor[D <: WorldObject, R <: WorldObject] extends Processor {

  def customCommonExpression[D1 <: D, R1 <: R](d: D1, r: R1): Relation[D1, R1]

  def commonExpression[D1 <: D, R1 <: R](d: D1, r: R1): Relation[D1, R1] = (d, r) match {
    case (Nihility, _) | (_, Nihility) => new Relation[D1, R1](d, r)
    case (_, _) => customCommonExpression(d, r)
  }

}
